src/vscode-bicep/src/panes/deploy/app/components/localDeploy.tsx

```diff
@@ -0,0 +1,76 @@
+import { VSCodeDataGrid, VSCodeDataGridCell, VSCodeDataGridRow } from "@vscode/webview-ui-toolkit/react";
+import { FormSection } from "./sections/FormSection";
+import { LocalDeployResponse, LocalDeploymentOperationContent } from "../../../../language";
+import { getPreformattedJson } from "./utils";
+
+export function getLocalDeployResult(result: LocalDeployResponse) {
+  const error = result.deployment.error;
+  return (
+    <FormSection title="Result">
+      <p>{result.deployment.provisioningState}</p>
+      {error && 
+        <VSCodeDataGrid gridTemplateColumns="max-content auto">
+        {error && <VSCodeDataGridRow key={0}>
+          <VSCodeDataGridCell gridColumn="1">Code</VSCodeDataGridCell>
+          <VSCodeDataGridCell gridColumn="2">{error.code}</VSCodeDataGridCell>
+        </VSCodeDataGridRow>}
+        {error.message && <VSCodeDataGridRow key={1}>
+          <VSCodeDataGridCell gridColumn="1">Message</VSCodeDataGridCell>
+          <VSCodeDataGridCell gridColumn="2">{error.message}</VSCodeDataGridCell>
+        </VSCodeDataGridRow>}
+      </VSCodeDataGrid>}
+    </FormSection>
+  );
+}
+
+export function getLocalDeployOperations(result: LocalDeployResponse) {
+  if (!result.operations) {
+    return null;
+  }
+
+  return (
+    <FormSection title="Operations">
+      <VSCodeDataGrid>
+        <VSCodeDataGridRow rowType="header">
+          <VSCodeDataGridCell gridColumn="1" cellType="columnheader">Resource Name</VSCodeDataGridCell>
+          <VSCodeDataGridCell gridColumn="2" cellType="columnheader">State</VSCodeDataGridCell>
+          <VSCodeDataGridCell gridColumn="3" cellType="columnheader">Error</VSCodeDataGridCell>
+        </VSCodeDataGridRow>
+        {result.operations.map(operation => (
+          <VSCodeDataGridRow key={operation.resourceName} style={isFailed(operation) ? { background: "rgba(255, 72, 45, 0.3)" } : {}}>
+            <VSCodeDataGridCell gridColumn="1">{operation.resourceName}</VSCodeDataGridCell>
+            <VSCodeDataGridCell gridColumn="2">{operation.provisioningState}</VSCodeDataGridCell>
+            <VSCodeDataGridCell gridColumn="3">{operation.error ? getPreformattedJson(operation.error) : ''}</VSCodeDataGridCell>
+          </VSCodeDataGridRow>
+        ))}
+      </VSCodeDataGrid>
+    </FormSection>
+  );
+}
+
+export function getLocalDeployOutputs(result: LocalDeployResponse) {
```

Comment:
It might be worth changing these functions into components as a best practice to avoid unnecessary rendering.

Reply:
Pushed a change - is that what you were thinking?